### PR TITLE
Make debug info lazy

### PR DESCRIPTION
**Pull request by gaetschwartz**
# Description

* Makes the `debug info` lazy which greatly improves performance.
* Adds a `thread id` attribute

Screenshot 2023-10-15 211940
image

# User-Facing Changes

`threadid` column added.

### CODE REVIEW

**Comment by fdncred** [comment]
Do you have some performance/benchmark tests/metrics that empirically show that it's faster?

File: crates/nu-command/src/debug/info.rs [hunk 1]
```diff
@@ -34,25 +47,186 @@ impl Command for DebugInfo {
...
+                // get the process information for the nushell pid
+                let pinfo = system.process(pid);
+
+                if let Some(p) = pinfo {
+                    Ok(Value::record(
+                        record! {
+                            "memory" => Value::filesize(p.memory() as i64, self.span),
+                            "virtual_memory" => Value::filesize(p.virtual_memory() as i64, self.span),
+                            "status" => Value::string(p.status().to_string(), self.span),
+                            "root" => Value::string(p.root().to_string_lossy().to_string(), self.span),
```
Comment (fdncred) [inline crates/nu-command/src/debug/info.rs:146]: We cannot use `root()` as is right now; not until there's a new version `sysinfo` crate release. This is why I have the changes above getting "root" differently, with the comments.

File: crates/nu-command/src/debug/info.rs [hunk 2]
```diff
@@ -1,16 +1,29 @@
...
+    {
+        ';'
+    }
+    #[cfg(not(target_family = "windows"))]
+    {
+        ':'
+    }
+};
+
+macro_rules! equal_to_any {
```
Comment (fdncred) [inline crates/nu-command/src/debug/info.rs:19]: Would rather not have macros if possible.

File: crates/nu-command/src/debug/info.rs [hunk 3]
```diff
@@ -34,25 +47,186 @@ impl Command for DebugInfo {
...
+                                                }
+                                            }
+                                        };
+                                        if is_env_var_a_list {
+                                            let items = value.split(ENV_PATH_SEPARATOR_CHAR).map(|r| Value::string(r.to_string(), self.span)).collect::<Vec<_>>();
+                                            env_rec.push(key.to_string(), Value::list(items, self.span));
+                                            continue;
+                                        }
+
+                                        #[cfg(not(target_family = "windows"))]
```
Comment (fdncred) [inline crates/nu-command/src/debug/info.rs:173]: Why is windows excluded here?

File: crates/nu-command/src/debug/info.rs [hunk 4]
```diff
@@ -34,25 +47,186 @@ impl Command for DebugInfo {
...
-    ) -> Result<PipelineData, ShellError> {
-        let span = Span::unknown();
-        // get the nushell process id
+        column: &str,
+        system_option: Option<&System>,
+    ) -> Result<Value, ShellError> {
         let pid = Pid::from(std::process::id() as usize);
-        // only refresh the process and memory information
+        match column {
+            "threadid" => Ok(Value::int(get_thread_id(), self.span)),
```
Comment (fdncred) [inline crates/nu-command/src/debug/info.rs:86]: I'm getting `8153895680` as a thread_id on MacOS. Is that reasonable? I don't know, just seemed weird to me.
```suggestion
            "thread_id" => Ok(Value::int(get_thread_id(), self.span)),
```

Comment (gaetschwartz) [reply crates/nu-command/src/debug/info.rs:86]: This should be correct AFAIK, phthread_self returns the current thread id in libc: https://opensource.apple.com/source/Libc/Libc-583/pthreads/pthread_self.3.auto.html

File: crates/nu-command/src/debug/info.rs (hunk 1 above)
Comment (gaetschwartz) [reply crates/nu-command/src/debug/info.rs:146]: Nice catch, missed this while copy-pasting !

File: crates/nu-command/src/debug/info.rs (hunk 3 above)
Comment (gaetschwartz) [reply crates/nu-command/src/debug/info.rs:173]: LS_COLORS is Unix only so I figured its best to ignore it. Though now that I can remove this.

File: crates/nu-command/src/debug/info.rs (hunk 2 above)
Comment (gaetschwartz) [reply crates/nu-command/src/debug/info.rs:19]: Too bad, that was a pretty clean way of checking different variables for different configs !

File: crates/nu-command/src/debug/info.rs (hunk 3 above)
Comment (fdncred) [reply crates/nu-command/src/debug/info.rs]: Thanks. FYI - You can use LS_COLORS on windows with nushell. One popular way is to do something like `$env.LS_COLORS = (vivid generate catppuccin-mocha)`. This allows you to theme your `ls` names.

File: crates/nu-command/src/debug/info.rs (hunk 4 above)
Comment (fdncred) [reply crates/nu-command/src/debug/info.rs]: Thanks. Just looked weird to have such a large number as a thread_id.

**Comment by fdncred** [comment]
We're too close right now to a new nushell release. Let's wait until after the release and we can land this. Thanks for the screenshots/benchmarking and making the changes I suggested.

File: crates/nu-command/src/debug/info.rs (hunk 4 above)
Comment (gaetschwartz) [reply crates/nu-command/src/debug/info.rs]: Yeah AFAIK this isn't a monotonously increasing number, possibly just returns something like the pointer to the tread handle.

**Comment by gaetschwartz** [comment]
Note that I realized thread_id is not handled properly when compiled to wasm, but AFAIK we don't plan to support such a platform anytime soon. Should I still provide some blanket implementation ?

**Comment by fdncred** [comment]
I think we're fine without wasm right now. We used to support wasm but it's been a long while now.

**Comment by gaetschwartz** [comment]
Oh it is in the screenshots but I'm not sure I kept them :P

**Comment by sholderbach** [comment]
A lazy solution I slapped together for nushell changes that are just a linear trail of commits going back: https://gist.github.com/sholderbach/280a04e80e57948a6d94670dacbd2e6e Runs a script you provide to compute the metric you care about.

**Comment by gaetschwartz** [comment]
I think this PR is ready to be merged, thoughts ? @fdncred @sholderbach

**Comment by fdncred** [comment]
Something I'm still confused about. The slowest part of this process has always been the `get_system()` call with the `RefreshKind`. It seems like it's called multiple times now instead of just once, which should make this slower. What am I missing?

**Comment by gaetschwartz** [comment]
Exactly, so when I collect I provide a precomputed system so it's not slower, but we gain a lot when querying single columns ;)

**Comment by fdncred** [comment]
Thanks!

**Comment by fdncred** [comment]
Here's a couple screenshots just to show some of the things I was thinking of doing with `debug info`
image
change the formatting
image

hopefully everything is working and not being too lazy here. it's hard for me to tell.

**Comment by fdncred** [comment]
@gaetschwartz This lazy change doesn't seem to be working any longer https://github.com/nushell/nushell/issues/11901. Would you mind taking a look at it?

I can get the right results with the code below but not with the existing code.
```rust
                let mut system = System::new();
                // get the process information for the nushell pid
                system.refresh_processes_specifics(
                    ProcessRefreshKind::everything()
                        .without_cpu()
                        .without_disk_usage()
                        .without_user(),
                );
                let pinfo = system.process(pid);

```

I'm not sure what changed to break this.

**Pull request merged**